src/mango/test/eunit/mango_couch_views_indexer_test.erl

```diff
@@ -0,0 +1,193 @@
+-include_lib("fabric/test/fabric2_test.hrl").
+
+indexer_test_() ->
+    {
+        "Test indexing",
+        {
+            setup,
+            fun setup/0,
+            fun cleanup/1,
+            {
+                foreach,
+                fun foreach_setup/0,
+                fun foreach_teardown/1,
+                [
+                    ?TDEF_FE(index_docs),
+                    ?TDEF_FE(index_lots_of_docs, 10),
+                    ?TDEF_FE(index_can_recover_from_crash, 60)
+                ]
+            }
+        }
+    }.
+
+
+setup() ->
+    Ctx = test_util:start_couch([
+        fabric,
+        couch_jobs,
+        couch_views,
+        mango
+    ]),
+    Ctx.
+
+
+cleanup(Ctx) ->
+    test_util:stop_couch(Ctx).
+
+
+foreach_setup() ->
+    DbName = ?tempdb(),
+    {ok, Db} = fabric2_db:create(DbName, [{user_ctx, ?ADMIN_USER}]),
+    Db.
+
+
+foreach_teardown(Db) ->
+    meck:unload(),
+    ok = fabric2_db:delete(fabric2_db:name(Db), []).
+
+
+index_docs(Db) ->
+    DDoc = generate_docs(Db, 5),
+    wait_while_ddoc_builds(Db),
+    Docs = run_query(Db, DDoc),
+    ?assertEqual([
+        [{id, <<"1">>}, {value, 1}],
+        [{id, <<"2">>}, {value, 2}],
+        [{id, <<"3">>}, {value, 3}],
+        [{id, <<"4">>}, {value, 4}],
+        [{id, <<"5">>}, {value, 5}]
+    ], Docs).
+
+
+index_lots_of_docs(Db) ->
+    DDoc = generate_docs(Db, 150),
+    wait_while_ddoc_builds(Db),
+    Docs = run_query(Db, DDoc),
+    ?assertEqual(length(Docs), 150).
+
+
+index_can_recover_from_crash(Db) ->
+    meck:new(mango_eval, [passthrough]),
+    meck:expect(mango_eval, index_doc, fun (Indexes, Doc) ->
+        Id = Doc#doc.id,
+        case Id == <<"2">> of
+            true ->
+                meck:unload(mango_indexer),
+                throw({fake_crash, test_jobs_restart});
+            false ->
+                meck:passthrough([Indexes, Doc])
+        end
+    end),
+    DDoc = generate_docs(Db, 3),
+    wait_while_ddoc_builds(Db),
+    Docs = run_query(Db, DDoc),
+    ?assertEqual([
+        [{id, <<"1">>}, {value, 1}],
+        [{id, <<"2">>}, {value, 2}],
+        [{id, <<"3">>}, {value, 3}]
+    ], Docs).
+
+
+wait_while_ddoc_builds(Db) ->
+    fabric2_fdb:transactional(Db, fun(TxDb) ->
+        Ready = lists:filter(fun (Idx) ->
+            Idx#idx.build_status == ?INDEX_READY
+        end, mango_idx:list(TxDb)),
+
+        if length(Ready) > 1 -> ok; true ->
+            timer:sleep(100),
+            wait_while_ddoc_builds(Db)
+        end
+    end).
+
+
+run_query(Db, DDoc) ->
+    Args = #mrargs{
+        view_type = map,
+        reduce = false,
+        include_docs = true,
+        update = false
+    },
+    [Idx] = mango_idx:from_ddoc(Db, couch_doc:to_json_obj(DDoc, [])),
+    Cursor = #cursor{
+        db = Db,
+        index = Idx,
+        user_acc = []
+    },
+    Name = mango_idx:name(Idx),
+    CB = fun query_cb/2,
+    {ok, Cursor1} = couch_views:query(Db, DDoc, Name, CB, Cursor, Args),
```

Comment:
Seems fairly odd that we're not using mango to execute this query.

Reply:
I'm actually thinking these test should be changed to couch_views_indexer since its actually testing the couch_views_indexer and isn't really related to anything specificly mango

Reply:
Seems reasonable to me.